Src/LiquidProjections/ProjectionException.cs

```diff
@@ -0,0 +1,375 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.Serialization;
+using System.Text;
+using System.Threading;
+
+using LiquidProjections.Logging;
+
+using Newtonsoft.Json;
+
+namespace LiquidProjections
+{
+    /// <summary>
+    /// An exception describing an unrecoverable error in a projector.
+    /// </summary>
+    [Serializable]
+    public class ProjectionException : Exception
+    {
+        /// <summary>
+        /// Throws a new <see cref="ProjectionException"/> with the given inner exceptions
+        /// or enriches an existing <see cref="ProjectionException"/> among the <paramref name="innerException"/>
+        /// and its inner exceptions with additional data.
+        /// The enriched exception should be rethrown via "throw" without arguments.
+        /// </summary>
+        /// <param name="innerException">An already thrown exception.</param>
+        /// <param name="setProperties">
+        /// The delegate that enriches the provided <see cref="ProjectionException"/> with additional data.
+        /// </param>
+        public static void ThrowNewOrEnrichExisting(Exception innerException, Action<ProjectionException> setProperties)

```

Comment:
Why do you need this factory here instead of having the throwing party create an instance and throw it explicitly?